### PR TITLE
Alanren/integration test

### PR DESCRIPTION
**Pull request by alanrenmsft**
enable support for integration test:
Extension Test: Test the stuff that can not be tested with web ui testing. e.g. context menus, internal commands.
Smoke Test:  UI test

Updated the wiki to include how to run the tests: https://github.com/Microsoft/azuredatastudio/wiki/How-to-Contribute#automated-testing

### CODE REVIEW

File: extensions/integration-tests/src/testConfig.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	let servers = await getTestingServers();
+	return servers[0];
+}
+
+export async function getTestingServers(): Promise<TestServerProfile[]> {
+	let promise = new Promise<TestServerProfile[]>(resolve => {
+		resolve(TestingServers);
+	});
+	await promise;
+	return promise;
```
Comment (MattIrv) [inline extensions/integration-tests/src/testConfig.ts]: looks like this function should just return `TestingServers`, or `Promise.resolve(TestingServers)` if you need it to be a Promise. Or do you have plans to change it soon?

File: extensions/integration-tests/src/testConfig.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	public get providerDisplayName(): string { return getEnumMappingEntry(connectionProviderMapping, this.provider).displayName; }
+	public get authenticationType(): AuthenticationType { return this._profile.authenticationType; }
+	public get authenticationTypeName(): string { return getEnumMappingEntry(authenticationTypeMapping, this.authenticationType).name; }
+	public get authenticationTypeDisplayName(): string { return getEnumMappingEntry(authenticationTypeMapping, this.authenticationType).displayName; }
+}
+
+var TestingServers: TestServerProfile[] = [
+	new TestServerProfile(
+		{
+			serverName: 'SQLTOOLS2017-3',
```
Comment (MattIrv) [inline extensions/integration-tests/src/testConfig.ts]: Are you planning to make this configurable? The Tools Service integration tests use your vscode-mssql server config, so we could do something similar here

File: extensions/integration-tests/src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import * as sqlops from 'sqlops';
+import * as vscode from 'vscode';
+import { TestServerProfile } from './testConfig';
+
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+export async function waitForCompletion(thenable: Thenable<any>): Promise<any> {
```
Comment (MattIrv) [inline extensions/integration-tests/src/utils.ts]: is this function necessary? Based on the implementation it should just be the same as awaiting on the `thenable` directly

Comment (MattIrv) [reply extensions/integration-tests/src/utils.ts]: If it is necessary I'd suggest parameterizing it `export async function waitForCompletion<T>(thenable: Thenable<T>): Promise<T>` so that you don't need type assertions everywhere you call it

File: src/sql/parts/objectExplorer/common/objectExplorerService.ts [hunk 4]
```diff
@@ -77,6 +77,8 @@ export interface IObjectExplorerService {
 	getTreeNode(connectionId: string, nodePath: string): Thenable<TreeNode>;
 
 	refreshNodeInView(connectionId: string, nodePath: string): Thenable<TreeNode>;
+
+	getNodeActions(connectionId: string, nodePath: string): Thenable<string[]>;
```
Comment (MattIrv) [inline src/sql/parts/objectExplorer/common/objectExplorerService.ts]: Is this only added so that it can be tested? We should find another way to test it if so

File: src/sql/parts/objectExplorer/viewlet/serverTreeView.ts [hunk 5]
```diff
@@ -79,6 +81,14 @@ export class ServerTreeView {
 		return this._onSelectionOrFocusChange.event;
 	}
 
+	public get TreeActionProvider(): ServerTreeActionProvider {
+		return this._actionProvider;
+	}
+
+	public get Tree(): ITree {
```
Comment (MattIrv) [inline src/sql/parts/objectExplorer/viewlet/serverTreeView.ts]: this and the getter above should start with a lowercase letter for typescript

File: src/sql/sqlops.d.ts [hunk 6]
```diff
@@ -121,6 +121,12 @@ declare module 'sqlops' {
 	 * Namespace for interacting with Object Explorer
 	*/
 	export namespace objectexplorer {
+
+		/**
+		 * A testing hook
+		 */
+		export function getNodeActions(connectionId: string, nodePath: string): Thenable<string[]>;
```
Comment (MattIrv) [inline src/sql/sqlops.d.ts]: we shouldn't add public APIs just for our own testing

File: src/sql/sqlops.proposed.d.ts [hunk 7]
```diff
@@ -1209,7 +1209,7 @@ declare module 'sqlops' {
 			 * Registers a save handler for this editor. This will be called if [supportsSave](#ModelViewEditorOptions.supportsSave)
 			 * is set to true and the editor is marked as dirty
 			 */
-			registerSaveHandler(handler: () => Thenable<boolean>);
+			registerSaveHandler(handler: () => Thenable<boolean>): void;
```
Comment (MattIrv) [inline src/sql/sqlops.proposed.d.ts]: thanks for fixing these

File: src/vs/base/browser/ui/button/button.ts [hunk 8]
```diff
@@ -151,6 +151,9 @@ export class Button extends Disposable {
 			this.$el.addClass('monaco-text-button');
 		}
 		this.$el.text(value);
+		//{{SQL CARBON EDIT}}
+		this.$el.attr('aria-label', value);
+		//{{END}}
```
Comment (MattIrv) [inline src/vs/base/browser/ui/button/button.ts]: we have a ADS-specific button class at src/sql/base/browser/ui/button/button.ts, would it work to make this change there instead?

File: src/sql/workbench/common/actions.contribution.ts [hunk 9]
```diff
@@ -37,3 +37,16 @@ Registry.as<IConfigurationRegistry>(ConfigExtensions.Configuration).registerConf
 		}
 	}
 });
+
+Registry.as<IConfigurationRegistry>(ConfigExtensions.Configuration).registerConfiguration({
```
Comment (MattIrv) [inline src/sql/workbench/common/actions.contribution.ts]: why is this necessary? The connect dialog only gets shown on startup if there are no servers, generally first launch. I'm not sure of a case where we wouldn't want to show it

File: test/smoke/src/application.ts [hunk 10]
```diff
@@ -62,8 +62,15 @@ export class Application {
 
 	async start(): Promise<any> {
 		await this._start();
-		await this.code.waitForElement('.explorer-folders-view');
+		//{{SQL CARBON EDIT}}
+		await this.code.waitForElement('.object-explorer-view');
+
+		//Original
+		/*
+		await this.code.waitForElement('.explorer-folders-view');	 		await this.code.waitForElement('.object-explorer-view');
```
Comment (MattIrv) [inline test/smoke/src/application.ts]: looks like the line breaks got messed up here

File: test/smoke/test/index.js [hunk 11]
```diff
@@ -18,21 +18,22 @@ const opts = minimist(args, {
...
-	options.reporter = 'mocha-multi-reporters';
-	options.reporterOptions = {
-		reporterEnabled: 'spec, mocha-junit-reporter',
-		mochaJunitReporterReporterOptions: {
-			testsuitesTitle: `${suite} ${process.platform}`,
-			mochaFile: path.join(process.env.BUILD_ARTIFACTSTAGINGDIRECTORY, `test-results/${process.platform}-${suite.toLowerCase().replace(/[^\w]/g, '-')}-results.xml`)
-		}
-	};
-}
+//{{SQL CARBON EDIT}}
```
Comment (MattIrv) [inline test/smoke/test/index.js]: the original version of this is a special case for running in VSTS. Why don't we want to keep it here?

File: extensions/integration-tests/src/testConfig.ts (hunk 1 above)
Comment (alanrenmsft) [reply extensions/integration-tests/src/testConfig.ts]: we can't have user name/password saved in the code, in the near future we will need to get the list from other places to support sql authentication, e.g. Azure storage account. that's why I made this a Thenable.

File: extensions/integration-tests/src/testConfig.ts (hunk 2 above)
Comment (alanrenmsft) [reply extensions/integration-tests/src/testConfig.ts]: would like to know more detail about it.

File: extensions/integration-tests/src/utils.ts (hunk 3 above)
Comment (alanrenmsft) [reply extensions/integration-tests/src/utils.ts]: 👍 I didn't know you can await the Thenable, LOL. just tried it and this is not necessary.

File: src/sql/workbench/common/actions.contribution.ts (hunk 9 above)
Comment (alanrenmsft) [reply src/sql/workbench/common/actions.contribution.ts]: this is also for test environment setup, most test cases will start with no connection history (new temp user data folder). 1. for extension test (not UI automation based), there is no way for us to close the dialog programatically.
2. for ui automation, some test may not need this dialog to be on, and they have to close it, rather that having each test needing to worry about this, I would prefer ADS to start without this dialog opened.

File: test/smoke/src/application.ts (hunk 10 above)
Comment (alanrenmsft) [reply test/smoke/src/application.ts]: fixed

File: test/smoke/test/index.js (hunk 11 above)
Comment (alanrenmsft) [reply test/smoke/test/index.js]: yes, i am working on adding it back

File: src/vs/base/browser/ui/button/button.ts (hunk 8 above)
Comment (alanrenmsft) [reply src/vs/base/browser/ui/button/button.ts]: I can't override the setter, will cause a compilation error.

File: src/sql/sqlops.testapi.d.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+// This is the place for API experiments and proposal.
```
Comment (MattIrv) [inline src/sql/sqlops.testapi.d.ts]: this comment is out of date

File: src/sql/sqlops.testapi.d.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (MattIrv) [inline src/sql/sqlops.testapi.d.ts]: minor but i'd recommend calling it `sqlops.test.d.ts` instead

File: src/sql/parts/objectExplorer/common/objectExplorerService.ts (hunk 4 above)
Comment (MattIrv) [reply src/sql/parts/objectExplorer/common/objectExplorerService.ts]: can you at least add a tsdoc comment that this is just for testing so that we know it doesn't need to be implemented for anything else that implements this interface? probably good to make it optional too

File: src/sql/parts/objectExplorer/common/objectExplorerService.ts [hunk 14]
```diff
@@ -522,6 +524,14 @@ export class ObjectExplorerService implements IObjectExplorerService {
 		return Object.values(this._activeObjectExplorerNodes);
 	}
 
+	public getNodeActions(connectionId: string, nodePath: string): Thenable<string[]> {
```
Comment (MattIrv) [inline src/sql/parts/objectExplorer/common/objectExplorerService.ts]: likewise a comment here about what this does and why would be helpful

File: src/sql/sqlops.d.ts [hunk 15]
```diff
@@ -303,6 +304,13 @@ declare module 'sqlops' {
 		userName: string;
 	}
 
+	export interface ConnectionResult {
```
Comment (MattIrv) [inline src/sql/sqlops.d.ts]: this should be moved to `sqlops.proposed.d.ts` since it's only used there

File: src/sql/sqlops.proposed.d.ts [hunk 16]
```diff
@@ -1371,6 +1371,13 @@ declare module 'sqlops' {
 		 * @param callback
 		 */
 		export function openConnectionDialog(providers?: string[], initialConnectionProfile?: IConnectionProfile, connectionCompletionOptions?: IConnectionCompletionOptions): Thenable<connection.Connection>;
+
+		/**
+		 * Opens the connection and add it to object explorer and opens the dashboard. If connection was successful
+		 * returns the connection id otherwise returns undefined
```
Comment (MattIrv) [inline src/sql/sqlops.proposed.d.ts]: this comment doesn't match with the return type, can it be updated?

File: src/sql/sqlops.testapi.d.ts [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+ *--------------------------------------------------------------------------------------------*/
+
+// This is the place for API experiments and proposal.
+
+import * as core from 'sqlops';
+import * as vscode from 'vscode';
+
+declare module 'sqlops' {
+
+	export namespace extensionManagement {
```
Comment (MattIrv) [inline src/sql/sqlops.testapi.d.ts]: VS Code already calls their namespace `extensions` so let's use that name
